src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -368,78 +379,70 @@ public synchronized void removeShard(int shardId, String reason) throws Elastics
-                logger.debug("[{}] failed to close merge policy scheduler", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(MergePolicyProvider.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close merge policy provider", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(IndexShardGatewayService.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close index shard gateway", e, shardId);
-                // ignore
-            }
-            try {
-                // now we can close the translog
-                shardInjector.getInstance(Translog.class).close();
+                shardInjector.getInstance(Store.class).close();
             } catch (Throwable e) {
-                logger.debug("[{}] failed to close translog", e, shardId);
-                // ignore
+                logger.warn("[{}] failed to close store on shard deletion", e, shardId);
             }
+        }
+    }
+
+    private void close(ShardId shardId, Injector shardInjector, Class<? extends Closeable>... toClose) {
+        for (Class<? extends Closeable> closeable : toClose) {
             try {
-                // now we can close the translog
-                shardInjector.getInstance(PercolatorQueriesRegistry.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close PercolatorQueriesRegistry", e, shardId);
-                // ignore
+                final Closeable instance = shardInjector.getInstance(closeable);
+                assert instance != null;
+                IOUtils.close(instance);
+            } catch (Throwable t) {
+                logger.debug("{} failed to close {}", t, shardId, Strings.toUnderscoreCase(closeable.getSimpleName()));
             }
+        }
+    }
 
-            // call this before we close the store, so we can release resources for it
-            indicesLifecycle.afterIndexShardClosed(sId, indexShard);
-            // if we delete or have no gateway or the store is not persistent, clean the store...
-            final Store store = shardInjector.getInstance(Store.class);
-            // and close it
+    private void onShardClose(ShardLock lock) {
+        if (deleted.get()) { // we remove that shards content if this index has been deleted
             try {
-                    store.close();
-            } catch (Throwable e) {
-                logger.warn("[{}] failed to close store on shard deletion", e, shardId);
+                nodeEnv.deleteShardDirectorySafe(lock, indexSettings);
+            } catch (IOException e) {
+                logger.warn("{} failed to delete shard content", lock.getShardId());
             }
-            Injectors.close(injector);
+        }
+    }
+
+    private class StoreCloseListener implements Store.OnClose {
```

Comment:
Out of curiosity, why create a class for this instead of an anonymous class in `IndexService` capturing the local `shardId`? There are no other instantiations of this class other than the single one